fedimint-atomic-broadcast/src/broadcast.rs

```diff
@@ -1,155 +1,175 @@
-use std::time::Duration;
-
 use fedimint_core::db::Database;
-use fedimint_core::task::sleep;
 use fedimint_core::PeerId;
-use tokio::sync::{mpsc, oneshot, watch};
+use tokio::sync::{mpsc, oneshot};
+use tokio::task::JoinHandle;
 
 use crate::keychain::Keychain;
-use crate::{db, session, Decision, Message, OrderedItem, Recipient, Shutdown, SignedBlock};
+use crate::{db, session, Decision, Message, OrderedItem, Recipient, SignedBlock};
 
-async fn relay_messages(
+pub struct AtomicBroadcast {
+    keychain: Keychain,
     db: Database,
-    incoming_message_receiver: async_channel::Receiver<(Message, PeerId)>,
+    mempool_item_receiver: async_channel::Receiver<Vec<u8>>,
     outgoing_message_sender: async_channel::Sender<(Message, Recipient)>,
-    network_data_sender: async_channel::Sender<Vec<u8>>,
-    signed_block_sender: async_channel::Sender<SignedBlock>,
-) {
-    while let Ok((message, peer_id)) = incoming_message_receiver.recv().await {
-        match message {
-            Message::NetworkData(network_data) => {
-                network_data_sender.send(network_data).await.ok();
-            }
-            Message::BlockRequest(index) => {
-                if let Some(signed_block) = db::load_block(&db, index).await {
-                    outgoing_message_sender
-                        .send((Message::Block(signed_block), Recipient::Peer(peer_id)))
-                        .await
-                        .ok();
+    network_data_receiver: async_channel::Receiver<Vec<u8>>,
+    signed_block_receiver: async_channel::Receiver<SignedBlock>,
+    relay_handle: JoinHandle<()>,
```

Comment:
Task groups might help with managing this task too

Reply:
Probably, for the first broadcast implementation we once decided to go without them though.